idseq_dag/util/dict.py

```diff
@@ -14,18 +18,50 @@ class IdSeqDict(object):
         key has to be a string
         value could be either array of strings or a string
     '''
-    def __init__(self, db_path, value_type):
+    def __init__(self, db_path, value_type, read_only=True):
         self.db_path = db_path
         self.value_type = value_type
-        self.db_conn = sqlite3.connect(db_path, check_same_thread=False) # make it thread safe
-        cursor = self.db_conn.cursor()
-        cursor.execute(f"CREATE TABLE IF NOT EXISTS {SQLITE_TABLE_NAME} (dict_key VARCHAR(255) PRIMARY KEY, dict_value text)")
-        self.db_conn.commit()
+        self.read_only = read_only
+        self.uri_db_path = pathlib.Path(self.db_path).as_uri()
+        if self.read_only:
+            self.uri_db_path += "?mode=ro&immutable=1&nolock=1&cache=private"
+        with log.log_context(f"open db", {"db_path": self.db_path, "read_only": self.read_only, "uri_db_path": self.uri_db_path}):
+            self.db_conn = sqlite3.connect(self.uri_db_path, check_same_thread=False, uri=True) # make it thread safe
+        self._open = True
+        with log.log_context(f"assert_db_table", {"db_path": self.db_path, "read_only": self.read_only}):
+            cursor = self.db_conn.cursor()
+            if not self.read_only:
+                res = cursor.execute(f"SELECT count(*) FROM sqlite_master WHERE type='table' AND name='{SQLITE_TABLE_NAME}';")
+                v = res.fetchone()
+                if v[0] == 0:
+                    raise Exception(f"table {SQLITE_TABLE_NAME} doesn't exist in db {self.db_path}")
+            else:
+                cursor.execute(f"CREATE TABLE IF NOT EXISTS {SQLITE_TABLE_NAME} (dict_key VARCHAR(255) PRIMARY KEY, dict_value text)")
+                self.db_conn.commit()
+            cursor.close()
+
+    def __enter__(self):
```

Comment:
Is this and next function for logging purposes?

Reply:
No, the log part is the `log.log_context`.

This `__enter__` and `__exit__` functions are intended to make this `IDSeqDict` behave like  a context class. That allows us to do things like:
```
   with IDSeqDict(db_path, value_type) as db_conn:
       # ... do something with that connection
       # ... connection will be close immediately before exiting this block instead of waiting for the garbage collection (destructor behavior)
```

Reply:
I haven't replaced all places to use this new approach yet, but the class still works in the previous behavior though (open db and wait for garbage collector to close it).